vllm/engine/multiprocessing/mp_llm_engine.py

```diff
@@ -0,0 +1,259 @@
+                                          copy=False)
+
+    def run_engine_loop(self) -> None:
+        while True:
+            # Block until there is a new request.
+            if not self.engine.has_unfinished_requests():
+                self.wait_for_new_input()
+
+            # Handle any new input from the input socket.
+            self.maybe_handle_new_input()
+
+            # Engine step.
+            request_outputs = self.engine.step()
+
+            # Stream results to output socket.
+            self.stream_outputs(request_outputs)
+
+    def wait_for_new_input(self):
+        while self.input_socket.poll(timeout=POLLING_TIMEOUT_MS) == 0:
+            logger.debug("Waiting for new request.")
+
+    def stream_outputs(self, request_outputs: List[RequestOutput]):
+        self.output_socket.send_multipart((pickle.dumps(request_outputs), ),
+                                          copy=False)
+
+    def ack_check_health(self):
+        self.health_socket.send_multipart(
+            (pickle.dumps(VLLM_RPC_SUCCESS_STR), ), copy=False)
+
+    def maybe_handle_new_input(self):
+        """Handle new input with non-blocking IO"""
+        while self.input_socket.poll(timeout=0) != 0:
+            message = self.input_socket.recv(copy=False)
+            request = cloudpickle.loads(message.buffer)
+
+            if isinstance(request, RPCGenerateRequest):
+                self._handle_generate_request(request)
+            elif isinstance(request, RPCAbortRequest):
+                self._handle_abort_request(request)
+            elif isinstance(request, RPCUtilityRequest):
+                self._handle_utility_request(request)
+            else:
+                raise ValueError(f"Unknown RPCRequest: {request}")
+
+    def _handle_generate_request(self, request: RPCGenerateRequest):
+        self.engine.add_request(
+            request_id=request.request_id,
+            inputs=request.inputs,
+            params=request.sampling_params,
+            lora_request=request.lora_request,
+            trace_headers=request.trace_headers,
+            prompt_adapter_request=request.prompt_adapter_request,
+        )
+
+    def _handle_abort_request(self, request: RPCAbortRequest):
+        self.engine.abort_request([request.request_id])
+
+    def _handle_utility_request(self, request: RPCUtilityRequest):
+        if request == RPCUtilityRequest.DO_LOG_STATS:
+            self.engine.do_log_stats()
+        elif request == RPCUtilityRequest.CHECK_HEALTH:
+            self.engine.check_health()
+            self.ack_check_health()
+
+
+def run_mp_engine(engine_args: AsyncEngineArgs, usage_context: UsageContext,
+                  ipc_path: str):
+    engine = MPLLMEngine.from_engine_args(engine_args=engine_args,
+                                          usage_context=usage_context,
+                                          ipc_path=ipc_path)
+
+    engine.run_background_loop()
```

Comment:
As I understand from the code, the loop that it is running here is not actually run in "the background".

Reply:
`mp_llm_engine` runs in the background